### DISCUSSION TITLE
Global middleware with API Routing

### DISCUSSION THREAD

**Discussion by kennethacohen** [vercel/next.js#14415] [Help] [answered]
I'm currently working on converting my express custom server to API routes 🥳 

I had a couple of globally defined middlewares in my old express server, but am not sure how to handle in the world of API routes. Best solution I have so far is to create some kind of base middleware and use it in every single route, but it seems like a lot of duplicate code.

Any guidance on best practices would be greatly appreciated 😄

**Comment by trompx** [comment]
Hey @kennethacohen,

I am also wondering what are the best practice to implement global middlewares. This has been further discussed in https://github.com/vercel/next.js/issues/7208 but it seems it has been abandonned for now:

> It's unlikely this will land anytime soon as it introduces quite a bit of additional complexity for no additional gain based on upcoming features.

I just finished reading https://hoangvvo.com/blog/middlewares-in-next-js-without-custom-server/ which brings some light on the subject, so I may consider going the next-connect route.

**Comment by meoyawn** [comment, marked as answer]
`next-connect` is great but I would love a way to plug in my logger globally for all routes

In fact Next.js already has default global middleware built in https://nextjs.org/docs/api-routes/api-middlewares they're just not configurable

It would be so nice to have them configurable both on global and folder level (e.g. plug something just for `/api/auth`)

context: I'm building a complete website rails-style in Next.js

**Reply by kahur** [reply]
Well, if you still looking for a middle-ware plug & play. I've had the same and similar issue. So I've created my own middleware to handle other middleware. 

Here's the service:

```typescript
"use server";

import { NextFetchEvent, NextMiddleware, NextRequest, NextResponse } from "next/server";
import { MiddlewareRegistryType, MiddlewareType } from "./types";

const middleWareRegistry = (paths: string[], exclude = false) => {
  let _middlewares: MiddlewareType[] = [];

  const middlewares = (midlewareItems: MiddlewareType[]) => {
    _middlewares = midlewareItems;
  };

  const getMiddlewares = () => {
    return _middlewares;
  };

  const replacePathWithRegex = (path: string) => {
    return path.replaceAll("*", "([A-Za-z-_]+)");
  };

  const isMatching = (request: NextRequest) => {
    const pathname = request.nextUrl.pathname;
    const matching = paths.find(path => {
      let test = false;
      if (path.includes("*")) {
        test = new RegExp(replacePathWithRegex(path), "g").test(pathname);
      }

      if (test) {
        return true;
      }

      if (pathname.startsWith(path)) {
        return true;
      }

      return false;
    });

    if (exclude) {
      return matching ? false : true;
    }

    return matching ? true : false;
  };

  return { middlewares, getMiddlewares, isMatching };
};

/**
 * Can be used also with api routes for custom middlewares
 *
 * @param middlewares
 * @param index
 */
export const handleMiddlewares = (middlewares: MiddlewareType[] = [], index = 0) => {
  const current = middlewares[index] ?? null;
  if (current && typeof current === "function") {
    const next: NextMiddleware = handleMiddlewares(middlewares, index + 1);

    return current(next);
  }

  return request =>
    NextResponse.next({
      request: request,
    });
};

/**
 * middleware service managing middlewares and their executions and rules
 */
export const middlewareService = () => {
  const registry: MiddlewareRegistryType[] = [];

  const set = (paths: string[], exclude = false) => {
    const reg = middleWareRegistry(paths, exclude);
    registry.push(reg);

    return reg;
  };

  const execute = (request: NextRequest, event: NextFetchEvent) => {
    const middlewares = registry.flatMap(value => {
      if (value.isMatching(request)) {
        return value.getMiddlewares();
      }

      return [];
    });

    return handleMiddlewares(middlewares)(request, event);
  };

  return { set, execute };
};

```

here how you can use it inside of middleware.ts ( the global middleware ), where you can register middleware and set routes to be ignored / listen to. The boolean as second parameters is setting exluce or include
```typescript
export default function middleware(request: NextRequestWithAuth, event: NextFetchEvent) {
  const middlewareRegistryService = middlewareService();
middlewareRegistryService
      .set(["/onboarding", "/*/onboarding", "/api/auth", "/*/login", "/login"], true)
      .middlewares([withTokenAuthMiddleware]);

  // exclude pages from locales
  middlewareRegistryService.set(["/api"], true).middlewares([withLocaleMiddleware]);
  }
```

And than as last you can create set of middleware you want to use what gives you pretty much control over each middleware and routes you want them to be executed or ignored. 

Middleware than looks like this, for example with token middleware

```typescript
import { NextRequestWithAuth, withAuth } from "next-auth/middleware";
import { MiddlewareType } from "../types";
import { NextMiddlewareResult } from "next/dist/server/web/types";
import { URLS } from "@shared/utils/urls";

export const withTokenAuthMiddleware: MiddlewareType = next => {
  return async (request: NextRequestWithAuth, _next) => {
    const auth = await withAuth(request, {
      callbacks: {
        authorized: ({ token }) => {
          return !!token;
        },
      },
      pages: {
        signIn: URLS.login,
      },
    });

    if (auth) {
      return auth as NextMiddlewareResult;
    }

    return next(request, _next);
  };
};
```

here types: 

```typescript
import { NextMiddleware, NextRequest } from "next/server";

export interface MiddlewareRegistryType {
  middlewares: (middlewares: MiddlewareType[]) => void;
  getMiddlewares: () => MiddlewareType[];
  isMatching: (request: NextRequest) => boolean;
}

export type MiddlewareType = ((middleware: NextMiddleware) => NextMiddleware) | undefined;

```

Quite late reply. But hop it will help someone to solve this missing part in next.

**Comment by mattcarlotta** [comment]
Unfortunately, the current approach is to manually apply your own middlewares for routes. There was an RFC for global API middleware, but that was dropped as Vercel is seemingly more focused toward serverless setups. That said, you can still keep your express setup (check the [examples](https://github.com/vercel/next.js/tree/canary/examples) provided in the github repo for a custom express server), but it'll result in a slower TTFB (since you're technically running two servers). 

A work around to not using a custom express server would be to create a reusable middleware function that utilizes some express middlewares:

**middlewares/index.js**
```js
import morgan from "morgan";
import session from "cookie-session";
import passport from "passport";

const {
  cookieKey,
  DOMAIN,
  inDevelopment,
  inProduction,
  inStaging,
} = process.env;

export default next => async (req, res) => {
  try {
    // creates a list of middlewares (not required, but also filters any conditional 
    // middlewares based upon current ENV)
    const middlewares = [
      session({
        path: "/",
        keys: [cookieKey],
        name: "app",
        maxAge: 2592000000,
        httpOnly: true,
        domain: DOMAIN,
        secure: inProduction && !inStaging,
        sameSite: inProduction && !inStaging,
      }),
      inDevelopment && morgan("tiny"),
      passport.initialize(),
    ].filter(Boolean);
                 
    // each middleware will then be wrapped within its own promise
    const promises = middlewares.reduce((acc, middleware) => {
      const promise = new Promise((resolve, reject) => {
        middleware(req, res, result =>
          result instanceof Error ? reject(result) : resolve(result),
        );
      });
      return [...acc, promise];
    }, []);

    // promised middlewares get asynchronously resolved (this may need to be switched to a synchronous 
    // loop if a certain middleware function needs to be resolved before another)
    await Promise.all(promises);

    // returns the next wrapped function(s) to be executed (can be an API route or another additional middleware)
    return next(req, res);
  } catch (error) {
    // if any middleware fails, throws a 400 error
    return res.status(400).send(error);
  }
};
```

Then you can import the function above and wrap the default export of an API route:
**pages/api/example/index.js**
```js
import withMiddleware from "../../../middleware";

const handler = (req, res) => {
  // API route logic
};

// wrap the default export with the middleware function...
// the middleware function is executed first, then the handler 
// is executed "next"
export default withMiddleware(handler);
```

I find this to be the cleanest way to handle API middleware as it similar to express, but allows you to conditionally apply/chain middlewares without any additional 3rd party dependencies. If you need to add route-specific middleware, then you can chain it:

**middleware/requireAuth.js**
```js
const requireAuth = next => (req,res) => {
   // additional logic here

   return next(req, res);
};

export default requireAuth;
```

Then in your API route:
```js
import withMiddleware from "../../../middleware";
import requireAuth from "../../../middleware/requireAuth";

const handler = (req, res) => {
  // API route logic
};

// chain the default export with additional middlewares
export default withMiddleware(requireAuth(handler));
```

**Reply by kennethacohen** [reply]
yep, I landed in a similar place!

**Comment by Mihai-github** [comment]
Hey guys,
   
    A bit late to the party... but I'm curious if in the past years since this post originally was made if something has changed for this to be somehow available without not coming up with some sort of custom solution...
    Checking the docs seems like not but I'm curious about what you think :)

Thank you!
